### PR TITLE
[SPARK-26718][SS][BRANCH-2.4] Fixed integer overflow in SS kafka rateLimit calculation

### PR DESCRIPTION
**Pull request by linehrr**
## What changes were proposed in this pull request?

Fix the integer overflow issue in rateLimit.

## How was this patch tested?

Pass the Jenkins with newly added UT for the possible case where integer could be overflowed.

### CODE REVIEW

**Comment by gaborgsomogyi** [comment]
Please check commit message requirements + I think this can be unit tested.

**Comment by dongjoon-hyun** [comment]
Hi, @linehrr . Thank you for your first contribution.
Could you make a PR to `master` branch first? We usually merge into `master` first for the future release and do backport from it.

**Comment by linehrr** [comment]
@dongjoon-hyun 
Added unit test, it fails before the code fix change and succeeded after the fix. 
I can't put this PR against the master, since master does not even have this class, maybe master is too behind or too ahead.

**Comment by dongjoon-hyun** [comment]
Thank you for the UT, @linehrr .
And, do you mean your UT pass in the master?
For the master, the patch might be different.

**Comment by linehrr** [comment]
@dongjoon-hyun 
in master, the whole class `sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchReader.scala` is not even there.

**Comment by dongjoon-hyun** [comment]
@linehrr .
Do you want to find this? You can use `git grep` to find the line you need.
- https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala#L234

**Comment by linehrr** [comment]
@dongjoon-hyun oh that's neat. I thought the whole class got deprecated in master, but it's just got renamed. I will do a PR in master as well. but do we want to backport this fix to 2.4.0 branch as well ?

**Comment by dongjoon-hyun** [comment]
For `branch-2.4`, you can keep this PR since the patch is not the same.

**Comment by dongjoon-hyun** [comment]
BTW, if you do `git grep`, there are one additional instance like this. Please check that together.

**Comment by linehrr** [comment]
@dongjoon-hyun hmm.. looks like KafkaSource.scala has the exact same code there... 
this is not good, we have duplicated code. and also I can't find good place like shared base trait or class to abstract that into a shared function... for now, I will just fix 2 places.

**Review by srowen — COMMENTED** [review]
Seems OK to me, just open a PR for master first, yeah

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala
Comment (HeartSaVioR) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala]: nit: let's remove unnecessary change.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala
Comment (HeartSaVioR) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala]: nit: There's actually no rule on this, but most of the cases JIRA title is used as a prefix (search "SPARK-" against `*Suite.scala`) so might be better to follow the case.

**Review by HeartSaVioR — COMMENTED** [review]
LGTM except some minor nits. DISCLAIMER: I suggested the logic so my review is just around new test case.

**Comment by HeartSaVioR** [comment]
Btw, I guess this patch add a new UT so please update the content of PR. 

And we may want to get some information on commit message: requiring to open JIRA for details might be annoying if we need to deal with multiple commits for triaging some bugs. Don't need to contain all the thing, but IMHO the mandatory information would be 1. the bug is related to Kafka source in spark-sql-kafka 2. the patch fixes the integer overflow issue in rateLimit. Current title and content of PR are not clear where the bug is.

**Comment by linehrr** [comment]
submitted this PR for master branch here: https://github.com/apache/spark/pull/23666, changed a bit on the PR title and reverted some unnecessary changes mentioned by @HeartSaVioR.

**Comment by dongjoon-hyun** [comment]
Please update this PR like the patch at `master` branch.

**Comment by dongjoon-hyun** [comment]
BTW, I just noticed that your commit authorship is unmatched to GitHub info.
```
ryne.yang <ryne.yang@acuityads.com>
```

It's not a problem, but you may want to be visible as GitHub ID here. :)
- https://github.com/apache/spark/pulse/daily

**Comment by linehrr** [comment]
@dongjoon-hyun 
done. 
my github username got messed up by inheriting the global username set in my company laptop LOL. it's ok, no need to fix it.

**Comment by dongjoon-hyun** [comment]
ok to test

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala
Comment (dongjoon-hyun) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala]: In the master, we didn't touch line 191 and 192.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala
Comment (dongjoon-hyun) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala]: Let's remove one extra empty line here. In master, we don't have this.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala
Comment (dongjoon-hyun) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala]: So, as @HeartSaVioR suggested, we should not add this line.

**Comment by dongjoon-hyun** [comment]
Also, ping @maropu since this is also related to `branch-2.3`.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, all.
Merged to branch-2.4.

**Comment by dongjoon-hyun** [comment]
@linehrr . Could you make a PR once more? I cannot merge this to `branch-2.3` because there is a conflict on `branch-2.3`, too.

**Comment by linehrr** [comment]
no worries, I will make one for 2.3 soon.

**Comment by linehrr** [comment]
@dongjoon-hyun seeing the same error mentioned here showing up in 2.3 when trying to build the project: http://apache-spark-developers-list.1001551.n3.nabble.com/VOTE-SPARK-2-3-2-RC3-td24421.html

```
[ERROR] missing or invalid dependency detected while loading class file 'MetricsSystem.class'.
Could not access term eclipse in package org,
because it (or its dependencies) are missing. Check your build definition for
missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
A full rebuild may help if 'MetricsSystem.class' was compiled against an incompatible version of org.
[ERROR] missing or invalid dependency detected while loading class file 'MetricsSystem.class'.
Could not access term jetty in value org.eclipse,
because it (or its dependencies) are missing. Check your build definition for
missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
A full rebuild may help if 'MetricsSystem.class' was compiled against an incompatible version of org.eclipse.
[WARNING] one warning found
[ERROR] two errors found
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  13.405 s
[INFO] Finished at: 2019-01-29T17:13:23-05:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:3.2.2:testCompile (scala-test-compile-first) on project spark-sql-kafka-0-10_2.11: Execution scala-test-compile-first of goal net.alchim31.maven:scala-maven-plugin:3.2.2:testCompile failed.: CompileFailed -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException

```

**Comment by srowen** [comment]
I've seen that a few times in the past and never figured it out. Same with a few others here. I think it's a build problem somewhere, not a Spark thing. Just try a clean build.

**Comment by maropu** [comment]
Thanks for this work, @linehrr! I'll wait for your pr in branch-2.3.

**Comment by linehrr** [comment]
@srowen even after doing `mvn clean test` it's still showing up. any suggestions ?

**Comment by srowen** [comment]
I forget what I had to do the few times this happened. It did seem to happen after switching branches. Maybe ensure you shut down zinc? or if that doesn't work, delete all the spark artifacts under ~/.m2/repository?

**Comment by linehrr** [comment]
@srowen strange, I don't have zinc server running and even after I purged whole .m2/repository still the same. anyway, I submitted the PR, I will delegate the build and test to jenkins. patch is almost the same as other branches, so it should be fine.

**Pull request closed without merging**
